README.md
Analytics Platform Concourse Webhook Dispatcher
===============================================

-   Free software: MIT license

## Features

Listens for GitHub organisational webhooks and routes them to specified concourse pipelines.

### Why?
The Concourse GitHub resource works on polling the GitHub API every 1 minute. As the number of pipelines grow you can
end up going over the GitHub API rate limit.

The GitHub resource does support being triggered by a webhook, but that would mean manually setting up a webhook
for each pipeline. Our pipelines are set up dynamically so this wouldn't work for us.

The dispatcher allows us to create 1 organisation webhook that points at this dispatcher and this dispatcher
dynamically routes to each pipeline.


### Major assumption

1. Our pipeline names match github repository names. This dispatcher assumes any repository has a pipeline in
the `CONCOURSE_TEAM` with exactly the same name.
2. That all pipelines are configured to allow the same [webhook token](https://concourse-ci.org/resources.html#resource-webhook-token)

## Usage

<table>
<thead>
<tr class="header">
<th>Variable</th>
<th>Default</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>DEBUG</td>
<td>False</td>
<td>False</td>
</tr>
<tr class="even">
<td>PORT</td>
<td>8000</td>
<td>Port to listen on</td>
</tr>
<tr class="odd">
<td>SECRET</td>
<td><strong>No Default</strong></td>
<td>Webhook Secret</td>
</tr>
<tr class="odd">
<td>CONCOURSE_BASE_URL</td>
<td><strong>No Default</strong></td>
<td>Base URL of your concourse instance; include trailing slash</td>
</tr>
<tr class="even">
<td>CONCOURSE_WEBHOOK_TOKEN</td>
<td><strong>No Default</strong></td>
<td>Webhook secret that concourse will expect as a query param</td>
</tr>
<tr class="even">
<td>CONCOURSE_TEAM</td>
<td>main</td>
<td>Concourse team name</td>
</tr>
<tr class="even">
<td>CONCOURSE_DEFAULT_RESOURCE</td>
<td>release</td>
<td>The concourse resource to trigger the check on in your pipeline</td>
</tr>
<tr class="even">
<td>DEFAULT_EVENT</td>
<td>release</td>
<td>The github event to trigger a pipeline check for</td>
</tr>
</tbody>
</table>

## Default behaviour

An event received by this dispatcher (usually a github org hook) will be,
if it does then the pipeline with the **same name** as the repo in `CONCOURSE_TEAM` (default: `main`) will be 
triggered, the resource that will be triggered will be `DEFAULT_RESOURCE` (default: `release`)

## Running in docker

```bash
$ docker build . -t webhook-dispatcher
$ docker run --rm -it webhook-dispatcher
```

The Docker image is automatically built in [quay.io](https://quay.io) at https://quay.io/repository/mojanalytics/webhook-dispatcher/


## Running locally
make sure your working directory is the root of the checkout of this repo and run the following:
```bash
export PYTHONPATH="${PYTHONPATH}:${PWD}"
export CONCOURSE_BASE_URL=https://concourse.yourserver.com
export CONCOURSE_TEAM=main
export CONCOURSE_WEBHOOK_TOKEN="webhook-token-for-concourse-pipeline" # https://concourse-ci.org/resources.html#resource-webhook-token
export SECRET="your-github-webhook-secret" # https://developer.github.com/webhooks/securing/
python analytics_platform_concourse_webhook_dispatcher/cli.py server
```
You'll see the following output
``` 
[2018-11-01 15:44:36 +0000] [4014] [INFO] Goin' Fast @ http://0.0.0.0:8000
[2018-11-01 15:44:36 +0000] [4014] [INFO] Starting worker [4014]
```
at this point you have the dispatcher running but github can't call it, so you'll
have to expose it using something like ngrok, localtunnel or burrow. 
```bash
ngrok http 8000
ngrok by @inconshreveable                                                                                                             (Ctrl+C to quit)
                                                                                                                                                      
Session Status                online                                                                                                                  
Account                       you@yourdomain.com (Plan: Free)                                                                                
Version                       2.2.8                                                                                                                   
Region                        United States (us)                                                                                                      
Web Interface                 http://127.0.0.1:4040                                                                                                   
Forwarding                    http://5977ca2d.ngrok.io -> localhost:8000                                                                              
Forwarding                    https://5977ca2d.ngrok.io -> localhost:8000                                                                             
                                                                                                                                                      
Connections                   ttl     opn     rt1     rt5     p50     p90                                                                             
                              0       0       0.00    0.00    0.00    0.00                                                                            
                                                                             
```
Then create an webhook for your
organisation on github pointing to the `ngrok` url: https://5977ca2d.ngrok.io.

Your local server should now receive webhooks from github and forward them to the specified concourse server.

## Running tests
Follow the [Running locally] instructions
```bash
pip install -r requirements_dev.txt
export CONCOURSE_BASE_URL=https://httpbin.org
pytest
```

## Deploying to Kubernetes
Use this [helm chart](https://github.com/ministryofjustice/analytics-platform-helm-charts/tree/master/charts/webhook-dispatcher)

## Credits

This package was created with [Cookiecutter] and the
[audreyr/cookiecutter-pypackage] project template.

  [Cookiecutter]: https://github.com/audreyr/cookiecutter
  [audreyr/cookiecutter-pypackage]: https://github.com/audreyr/cookiecutter-pypackage


CHANGELOG.md
Changelog
=======

0.1.0 (2018-10-30)
------------------

* First version

0.1.1 (2018-10-31)
------------------
Updated: Bugfix when closing the server, it used to warn
about the event loop already being closed.

0.1.2 (2019-03-01)
------------------
Updated: Use python-fly instead of calling the web URL for concourse


tests/test_server.py
#!/usr/bin/env python

"""Tests for `analytics_platform_concourse_webhook_dispatcher` package."""
import os
import json
import re
from typing import MutableMapping
from unittest.mock import patch, Mock

import pytest


from analytics_platform_concourse_webhook_dispatcher.signature import make_digest

DIR = os.path.dirname(__file__)
CONCOURSE_BASE_URL_PATTERN = re.compile(r"https://httpbin\.org/.*$")


def sign(app, data):
    if data is None:
        data = b""
    if isinstance(data, MutableMapping):
        data = bytes(json.dumps(data), encoding="utf8")

    return make_digest(bytes(app.config.SECRET, encoding="utf8"), data)


@pytest.fixture
def app():
    from analytics_platform_concourse_webhook_dispatcher.server import app

    return app


def test_home(app):
    request, response = app.test_client.get("/")
    assert response.status == 200
    assert response.json == {}


def test_ping_unsigned(app):
    """
    test ping without signature, should fail
    """
    request, response = app.test_client.post(
        "/", headers={"X-Github-Event": "ping"})
    assert response.status == 400


def test_ping(app):
    digest = sign(app, None)

    request, response = app.test_client.post(
        "/", headers={"X-Github-Event": "ping", "X-Hub-Signature": digest}
    )
    assert response.status == 200
    assert response.json == {"msg": "pong"}


def test_implicit_ping(app):
    digest = sign(app, None)

    request, response = app.test_client.post(
        "/", headers={"X-Github-Event": "ping", "X-Hub-Signature": digest}
    )
    assert response.status == 200
    assert response.json == {"msg": "pong"}


mock_get_fly = Mock()
mock_login = Mock()
mock_run = Mock()


@patch("fly.Fly.get_fly", mock_get_fly)
@patch("fly.Fly.login", mock_login)
@patch("fly.Fly.run", mock_run)
def test_release_event(app):
    with open(f"{DIR}/data/release.json", "r") as f:
        data = json.load(f)
    digest = sign(app, data)
    request, response = app.test_client.post(
        "/", json=data, headers={"X-Github-Event": "release", "X-Hub-Signature": digest}
    )
    assert response.status == 204
    mock_get_fly.assert_called_once_with()
    mock_login.assert_called_once_with("", "", "main")
    mock_run.assert_called_once_with(
        "check-resource", 
        "--resource", 
        "Hello-World/release"
    )


analytics_platform_concourse_webhook_dispatcher/server.py
import hmac
import asyncio

from sanic import Sanic
from sanic.exceptions import abort
from sanic.response import json
from sanic.log import logger
from fly import Fly

from analytics_platform_concourse_webhook_dispatcher.config import Config
from analytics_platform_concourse_webhook_dispatcher import events
from .signature import make_digest

app = Sanic()
app.config.from_object(Config)

fly_lock = asyncio.Lock()


async def dispatch(event: str, body: dict):
    """
    trigger the corresponding webhook in concourse and return 204
    """

    # try the default by seeing if event
    # is DEFAULT_EVENT and triggering a webhook on DEFAULT_TEAM
    route = events.get(app, event, body)
    if not route:
        return json({}, status=404)

    concourse_url = app.config.CONCOURSE_BASE_URL

    run_args = [
        "check-resource",
        "--resource", f"{route['pipeline']}/{route['resource']}"
    ]

    fly = Fly(
        concourse_url=concourse_url,
        target=app.config.CONCOURSE_TEAM
    )
    async with fly_lock:
        fly.get_fly()
        fly.login(
            app.config.CONCOURSE_MAIN_USERNAME,
            app.config.CONCOURSE_MAIN_PASSWORD,
            app.config.CONCOURSE_TEAM,
        )

    fly.run(*run_args)
    logger.info('Calling: fly %s' % ' '.join(run_args))

    return json(route, status=204)


@app.middleware("request")
async def validateHook(request):
    secret = app.config.get("SECRET")

    if request.method == "POST":
        digest = make_digest(bytes(secret, encoding="utf8"), request.body)
        hook_signature = request.headers.get("x-hub-signature", "")
        match = hmac.compare_digest(digest, hook_signature)

        if not match:
            abort(400, "Invalid signature")


@app.route("/")
async def home(request):
    return json({})


@app.route("/", methods=["POST"])
@app.route("/<path:path>", methods=["POST"])
async def hook(request, path=None):
    event = request.headers.get("X-GitHub-Event", "ping")
    logger.info(request)
    if event == "ping":
        return json({"msg": "pong"})
    else:
        return await dispatch(event, request.json)


def serve(routes=None):
    if routes:
        app.config.ROUTES_FILE = routes

    app.run(host="0.0.0.0", port=app.config.PORT, debug=app.config.DEBUG)
    loop = asyncio.get_event_loop()
    if not loop.is_closed():
        pending = asyncio.Task.all_tasks()
        loop.run_until_complete(asyncio.gather(*pending))


if __name__ == "__main__":
    serve()
